### DISCUSSION TITLE
Only works from CMD

### DISCUSSION THREAD

**Discussion by johnm92008** [nref/BrowseRouter#11] [General]
Thank you for creating the BrowseRouter app.

For me, on Windows 11, it only works from the command prompt:
BrowseRouter.exe mozilla.org opens FireFox 
BrowseRouter.exe youtube.com opens Chrome.

However, if I click on a shortcut on my desktop or any link in a different browser, it does not follow the rules in the config.ini file.  Perhaps I have misunderstand how the app is meant to work.  For example, I thought a link in any app or browser with a youtube.com url should open the browser specified in the config.ini file.

Here is my config.ini file:

; Default browser is first in list
; Use `{url}` to specify UWP app browser details
[browsers]
ff = C:\Program Files\Mozilla Firefox\firefox.exe
chrome = C:\Program Files\Google\Chrome\Application\chrome.exe
edge = C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe

; Url preferences.
; Only * is treated as a special character (wildcard).
; Matches are domain-only. Protocols and paths are ignored.
; Use "*.blah.com" for subdomains, not "*blah.com" as that would also match "abcblah.com".
[urls]
mozilla.org = ff
google.com = chrome
youtube.com = chrome
cnn.com = chrome

; ? symbols from both sides will match the whole URL (not only web site name)
?*teams.cdn.office.net/evergreen-assets/safelinks/1/atp-safelinks.html?url*atlassian.net*? = edge

; / symbols from both sides will match web site and path 
;/(?:[0-9]{1,3}\.){3}[0-9]{1,3}/ = ff
; Default case. Added automatically
; * = whatever

[sources]
;* - Notepad = chrome
;Slack | Test = chrome
;*Microsoft Teams* = ff
;*Microsoft Visual Studio* = ff

Here is the shortcut on my desktop which opens the default browser, edge, (first in the list) instead of chrome:
https://www.youtube.com/watch?v=kQ8DGP9p2LY
BrowseRouter https://www.youtube.com/watch?v=lY1B18K5u_0 also does not work to open the correct browser.

**Comment by nref** [comment]
Hey John, thanks for reporting this potential issue, and sorry for the trouble. I will read your report in detail, try to reproduce it, and get back to you ASAP!

**Reply by nref** [reply]
Quick question, did you ever run `BrowseRouter.exe --register` as admin?

**Reply by johnm92008** [reply]
Thank you for your quick response.
Yes, I had registered it as admin.  And I just did so now to be sure.   I also made BrowseRouter the default browser.  I specified in the config.ini file that the URL youtube.com = chrome.  The link opens in FireFox, which is the 1st browser defined in the config.ini file.  If I change the 1st browser to be Edge. The the youtube link opens in Edge.
Please let me know if I have misconfigured something.  
Thank you for your help.

**Reply by epiphanyplx** [reply]
I think that you need to do `*youtube.com*` instead - or at least `*.youtube.com*.`

I can reproduce the same issue otherwise - it goes to first browser because it's default and the URL you went to didn't match the pattern I believe.

**Reply by johnm92008** [reply]
Thank you, but youtube.com is specified in the config.ini file. I am using Win 11 Home edition, by the way.

**Reply by epiphanyplx** [reply]
I see now that asterisks don't play nice on github. 

I meant for it to read
`*youtube.com*`
or 
`*.youtube.com*`

**Comment by johnm92008** [comment]
> epiphanyplx - I made that change and it works now. Thank you.  Perhaps the config.ini file shown here on the Code tab should be updated to reflect this requirement.

Another question: should clicking the link while in another browser follow the config.ini rules?   For example, while in Firefox, I click a YouTube link (which I specified to go to Chrome) should it launch in Chrome?

**Comment by epiphanyplx** [comment]
Oh, I'm not sure, I'm just another user.

I could see the browser overriding that - if you're already in it why would
it send you to it's competitor?  :)

**Reply by johnm92008** [reply]
Yes, however, there are such browser extensions: https://add0n.com/open-in.html

**Reply by nref** [reply]
Thanks @epiphanyplx for helping him out. Glad we got it sorted out.

@johnm92008  sorry the README wasn't helpful. The Url preferences section did call this out, but while being technically precise it was hard to understand. Also we didn't include any examples of that.

I went ahead and updated the README: 

Before:

```ini
; Url preferences.
; Only * is treated as a special character (wildcard).
; Matches are domain-only. Protocols and paths are ignored.
; Use "*.blah.com" for subdomains, not "*blah.com" as that would also match "abcblah.com".
[urls]
google.com = chrome
visualstudio.com = edge
mozilla.org = ff
```

```ini
; Url preferences.
; - Only * is treated as a special character (wildcard).
; - Only domains are matched. Don't include protocols e.g. "https://" or paths e.g. "/some/path?query=value"
; - Beware that subdomains don't match automatically, e.g. "youtube.com = chrome" would not launch Chrome for "www.youtube.com"
;   For that reason, you'll often want a leading "*." e.g. "*.youtube.com". 
;   Note: Don't use "*youtube.com" as that would also match e.g. "notyoutube.com".
[urls]
*.google.com = chrome
*.youtube.com = chrome
*.visualstudio.com = edge
*.mozilla.org = ff
```

To provide a bit more detail on why "youtube.com" doesn't work: when you click a shortcut in Windows Explorer, BrowserRouter sees the full URL not just "youtube.com". 

Example from my machine (I'm using Edge):

`"C:\Users\doug\AppData\Local\BrowseRouter\2023-05-25.log"`
```log
5/25/2023 17:16:10 BrowseRouter: Attempting to launch "https://www.youtube.com/watch?v=kQ8DGP9p2LY" for "Program Manager"
5/25/2023 17:16:10 BrowseRouter: Found URL preference "*" => "ff" ("C:\Program Files\Mozilla Firefox\firefox.exe")
5/25/2023 17:16:10 BrowseRouter: Launching C:\Program Files\Mozilla Firefox\firefox.exe with args " https://www.youtube.com/watch?v=kQ8DGP9p2LY"
```

When you change it to "*.youtube.com" that matches

```log
5/25/2023 17:16:23 BrowseRouter: Attempting to launch "https://www.youtube.com/watch?v=kQ8DGP9p2LY" for "Program Manager"
5/25/2023 17:16:23 BrowseRouter: Found URL preference "*youtube.com" => "edge" ("C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe")
5/25/2023 17:16:23 BrowseRouter: Launching C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe with args " https://www.youtube.com/watch?v=kQ8DGP9p2LY"

```

**Reply by johnm92008** [reply]
Thank you.
